### PR TITLE
[Perf][FA4] Warm up dense (paged) FA4 attention kernels

### PR DESCRIPTION
**Pull request by sfc-gh-goliaro**
# [Perf][FA4] Warm up dense (paged) FA4 attention kernels

**Repo:** `vllm-project/vllm` · **Branch:** `fa4-dense-cutedsl-warmup`

## Summary
The CuTeDSL warmup path (`register_cutedsl_warmup_provider`) currently only warms the
FA4 **MLA prefill** kernel. Standard GQA models on the `FLASH_ATTN` backend
(`FlashAttentionForwardSm100`) therefore **JIT-compile the dense prefill specialization**
(`q_stage=2`, paged, `seqused_k`) on the first serving request — a large one-time latency
spike (~8 s) that `jit_monitor` explicitly flags as
`CuTeDSL JIT compilation during inference: FlashAttentionForwardSm100`.

This registers the dense FA4 backend (`FlashAttentionMetadataBuilder`) as a CuTeDSL warmup
provider (new `vllm/model_executor/warmup/fa4_dense_config.py`) and precompiles the dense
decode and prefill specializations at startup, mirroring the existing MLA-prefill provider
in `mla/prefill/flash_attn.py`.

- Restricted to the vLLM-compile path (`CompilationMode.VLLM_COMPILE`), where query and
  paged-KV tensors have the contiguous/paged strides the compiled kernel bakes. Under
  `enforce_eager`, tensors can be strided views, so warmup is skipped and eager falls back
  to stock in-serving JIT (unchanged).
- Emits fp8-descale variants when the KV cache is fp8.

## Dependency (why this is a draft)
Requires the paged compile-spec support in `vllm-flash-attention`:
**vllm-project/flash-attention#158** (`compile_flash_attn_varlen_func_from_specs`:
`seqused_k`/`page_table`/`k_stride`/`descale`). The cmake `GIT_TAG` here points at that
PR's commit; it must be updated to the **merged** flash-attn commit before this can build
in CI. Kept as a **draft** until #158 lands.

## Not a duplicate
- #46182 (merged) and #47451 (open) add the CuTeDSL warmup **infrastructure** and apply it
  to FA4 **MLA prefill** only. Neither touches the dense backend
  (`vllm/v1/attention/backends/flash_attn.py`) nor warms `FlashAttentionForwardSm100` for
  standard GQA models. This builds on the merged #46182 provider registry and fills that gap.
- #46167 adds the runtime *monitor* for post-warmup CuTeDSL compiles (the detector), not the
  dense fix. Searched open PRs/issues for dense-FA4 / `FlashAttentionForwardSm100` warmup —
  none found.

## Tests run + results
Hardware: 1× B200 (SM100), vLLM built from this branch. Throughput via
`vllm bench throughput --attention-backend FLASH_ATTN --dataset-name random
--input-len 1024 --output-len 1024 --num-prompts 256`, cold CuTeDSL cache, `VLLM_PLUGINS=""`.

| Model | Metric | Before (stock main) | After (this PR) | FlashInfer ref |
|---|---|--:|--:|--:|
| Llama-3.1-8B (bf16) | output tok/s | 19,935 | **54,957** (2.75×) | 61,320 |
| Llama-3.1-8B (bf16) | in-serving CuTeDSL JIT | 1 (~8 s spike) | **0** | 0 |
| Qwen3-VL-8B-Instruct-FP8 | output tok/s | 21,875 | **64,567** (2.95×) | 73,984 |
| Qwen3-VL-8B-Instruct-FP8 | in-serving CuTeDSL JIT | 1 | **0** | 0 |

- Single-pass FA4-vs-FlashInfer throughput gap: **3.0× → ~1.12×**.
- **`jit_monitor`:** 0 `CuTeDSL JIT compilation during inference` during serving (was 1, the
  dense prefill kernel).
- **Eager fallback:** `--enforce-eager` runs to completion (rc=0), warmup skipped — no regression.

## Model evaluation (output correctness)
Greedy generation (`temperature=0`) with `FLASH_ATTN` vs `FLASHINFER` on Llama-3.1-8B is
**byte-identical** across test prompts (e.g. "17 × 23" → "391", primes → "2, 3, 5, 7, 11").
Warmup only pre-populates the same `compile_key` the engine would JIT, so numerics are
unchanged.

## Residual warmed gap — out of scope for this PR (kernel, not warmup)
After cold-start is eliminated, a ~12% gap remains on decode-heavy workloads. Isolated
attention-kernel microbenchmarks (FA4 vs FlashInfer trtllm, identical paged inputs, outputs
validated to match, CUDA-event timed) attribute it entirely to the FA4 **decode** kernel;
**prefill is at parity or faster**:

| Phase | Scenario (heads 32/8, hd128, bf16, paged) | FA4 vs FlashInfer |
|---|---|--:|
| decode | batch 64, ctx 1024 | 1.55× slower |
| decode | batch 256, ctx 1024 | 1.59× slower |
| decode | batch 256, ctx 4096 | 1.38× slower |
| prefill | batch 1, seq 4096 | 0.80× (FA4 faster) |
| prefill | batch 4, seq 2048 | 0.97× (parity) |
| prefill | batch 8, seq 1024 | 1.02× (parity) |

CUDA graphs (which vLLM already uses for FA4 decode as FULL graphs) eliminate FA4's per-call
host overhead, so at serving batch sizes these pure-kernel ratios drive the end-to-end gap.
Closing it would require FA4 decode-kernel work (a dedicated `q_stage=1` kernel / better
split-KV vs FlashInfer's trtllm-gen decode cubin) — a separate effort, not this PR.

## Reviewer notes
- `ruff check` / `ruff format` clean on the changed files; run `pre-commit` before merging.
- Scope: this PR closes the **cold-start** gap only; the residual decode-kernel gap above is
  intentionally not addressed here.

🤖 AI-assisted change; a human must review every line and run the tests/evals before submitting.

### CODE REVIEW

File: vllm/model_executor/warmup/fa4_dense_config.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from collections.abc import Hashable, Iterator
+from dataclasses import dataclass
+
+import torch
+
+from vllm.model_executor.warmup.cutedsl_warmup import CuTeDSLCompileUnit
+
+FA4_DENSE_STANDARD_DTYPES = (torch.bfloat16, torch.float16)
+
+# A prefill-sized query (rows > q tile) selects the q_stage=2 kernel; a
```
Comment (LopezCastroRoberto) [inline vllm/model_executor/warmup/fa4_dense_config.py]: Since you’ve identified the lowest-level compile keys here already, could we use `q_stage` instead of higher-level variables such as prefill size? e.g., for FA4 MLA https://github.com/vllm-project/vllm/blob/68ea76e78005bb53d9ef6b855b6c23411f5bb1e1/vllm/model_executor/warmup/fa4_cutedsl_config.py#L147

In general, I think we should identify the atomic factors that determine different compile keys, and it looks like you've already identified most of them. I would use the atomic ones whenever possible.

File: vllm/model_executor/warmup/fa4_dense_config.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# single-token-per-seq query selects q_stage=1. Concrete magnitudes below the
+# tile boundary do not change the compiled kernel, only the q_stage selection.
+_PREFILL_Q_ROWS = 256
+_DECODE_BATCH = 8
```
Comment (LopezCastroRoberto) [inline vllm/model_executor/warmup/fa4_dense_config.py]: Same thing for this one, i.e., ideally we should try to find the atomic factors that trigger a different compile-key. Probably batch size is now the fundamental factor?

File: vllm/model_executor/warmup/fa4_dense_config.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    head_dim: int
+    page_size: int
+    max_blocks_per_seq: int
+    scale: float
+    fa_version: int
+    has_fp8_kv: bool = False
+
+
+def _supports_fa4_dense(ctx: FA4DenseCompileContext) -> bool:
+    major, _ = torch.cuda.get_device_capability()
```
Comment (LopezCastroRoberto) [inline vllm/model_executor/warmup/fa4_dense_config.py]: Ideally, this should use `current_platform`, not `torch.cuda` directly. See https://github.com/vllm-project/vllm/pull/46182/changes#r3501268810

File: vllm/model_executor/warmup/fa4_dense_config.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    return (
+        ctx.fa_version == 4
+        and ctx.dtype in FA4_DENSE_STANDARD_DTYPES
+        and ctx.num_qo_heads > 0
+        and ctx.num_kv_heads > 0
+        and ctx.head_dim > 0
+        and ctx.page_size > 0
+        and ctx.max_blocks_per_seq > 0
+        and major == 10
```
Comment (LopezCastroRoberto) [inline vllm/model_executor/warmup/fa4_dense_config.py]: What about SM90 and SM120? What are the compile-keys in that case?

**Review by LopezCastroRoberto — COMMENTED** [review]
Thanks for extending the FA4 warmup coverage. Great work, @sfc-gh-goliaro! 

Left some comments.

File: vllm/model_executor/warmup/fa4_dense_config.py (hunk 3 above)
Comment (sfc-gh-goliaro) [reply vllm/model_executor/warmup/fa4_dense_config.py]: Done! now using current_platform.get_device_capability()

File: vllm/model_executor/warmup/fa4_dense_config.py (hunk 4 above)
Comment (sfc-gh-goliaro) [reply vllm/model_executor/warmup/fa4_dense_config.py]: Added an arch-family helper. Per the FA4 selector: `q_stage=2` exists only on SM100/SM110 (SM90 & SM120 are always `q_stage=1`); split-KV is rejected on SM120; fp8 is SM100-only. So we emit: SM90 `{q1}x{split,no-split}`, SM100/110 `{q1,q2}x{split,no-split}`, SM120 `{q1}`. Unknown arches skip warmup (fall back to JIT).

File: vllm/model_executor/warmup/fa4_dense_config.py (hunk 1 above)
Comment (sfc-gh-goliaro) [reply vllm/model_executor/warmup/fa4_dense_config.py]: Done. Now driven by the atomic factor: `q_stage=2` via `max_seqlen_q = tile_m + 1` (SM100 tile 128), `q_stage=1` via `max_seqlen_q=1`. This is the same pattern as the MLA config.

File: vllm/model_executor/warmup/fa4_dense_config.py (hunk 2 above)
Comment (sfc-gh-goliaro) [reply vllm/model_executor/warmup/fa4_dense_config.py]: > Probably batch size is now the fundamental factor?

Hmm not batch, batch isn't in compile_key. The atomic decode factor is `is_split_kv` (`num_splits>1`); batch only flips it through FA4's split heuristic, and serving passes `num_splits=0`. Now enumerated directly via `num_splits` $\in$ `{1,2}` , which also adds the split-KV decode kernel the old `num_splits=1`-only version missed.

**Comment by sfc-gh-goliaro** [comment]
Merged the changes from main and addressed @LopezCastroRoberto's comments: the warmup now enumerates atomic compile_key factors (`q_stage`, `is_split_kv`) directly, uses current_platform, and covers all FA4 arch families (SM90/SM100-110/SM120).

Dropped the fp8/descale branch: the FA4 fp8 kernel writes a bf16 output, but `compile_flash_attn_varlen_func_from_specs` hardcodes the compile-only output `dtype` to `q_dtype` (no `out_dtype`), so an fp8-in/bf16-out spec can't be built. fp8-KV dense serving falls back to in-serving JIT until that helper gains an out_dtype.

Validated on B200 (SM100): all 4 bf16 dense specs JIT-compile and select the intended `q_stage`/`is_split_kv`/`paged`/`seqused_k` kernels. 

Still draft as cmake pin stays at the https://github.com/vllm-project/flash-attention/pull/158 commit until it lands in main.

**Review by LopezCastroRoberto — COMMENTED** [review]
Thanks a lot for doing this, @sfc-gh-goliaro!

Would you mind moving this to the warmup contract we're trying to establish? See: https://github.com/vllm-project/vllm/pull/47451. Specifically `vllm/v1/attention/backends/mla/prefill/flash_attn.py`. RFC here: https://github.com/vllm-project/vllm/issues/47456

I think it should be very straightforward in your case, since you identified the core compile-keys already.
